diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/classes/SingleConstructorRule.kt

```diff
@@ -0,0 +1,151 @@
+            .map {
+                // non-null assert is safe because of predicate in partitioning
+                (it as KtBinaryExpression).asAssignment()!!.left!!
+            }
+            .filter {
+                // todo use secondaryCtor.findAllVariablesWithAssignments()
+                (it as? KtDotQualifiedExpression)?.run {
+                    receiverExpression is KtThisExpression && selectorExpression is KtNameReferenceExpression
+                } ?: false ||
+                        it is KtNameReferenceExpression
+            }
+            .partition { it is KtDotQualifiedExpression }
+            .let { (qualified, simple) ->
+                qualified.map { (it as KtDotQualifiedExpression).selectorExpression as KtNameReferenceExpression } to
+                        simple.map { it as KtNameReferenceExpression }
+            }
+
+        val assignedClassProperties = qualifiedAssignedProperties + assignedProperties
+            .run {
+                val localProperties = secondaryCtor.findAllNodesWithSpecificType(ElementType.PROPERTY)
+                filterNot { reference ->
+                    // check for shadowing
+                    localProperties.any {
+                        reference.node.isGoingAfter(it) && (it.psi as KtProperty).name == reference.name
+                    }
+                }
+            }
+
+        // find properties declarations
+        val declarationsAssignedInCtor = assignedClassProperties
+            .mapNotNull { reference ->
+                (classNode.psi as KtClass).getProperties()
+                    .firstOrNull { it.nameIdentifier?.text == reference.getReferencedName() }
+            }
+            .distinct()
+
+        classNode.convertSecondaryConstructorToPrimary(secondaryCtor, declarationsAssignedInCtor, otherStatements)
+    }
+
+    @Suppress("UnsafeCallOnNullableType", "NestedBlockDepth")
+    private fun ASTNode.convertSecondaryConstructorToPrimary(secondaryCtor: ASTNode,
+                                                             declarationsAssignedInCtor: List<KtProperty>,
+                                                             otherStatements: List<KtExpression>) {
+        require(elementType == CLASS)
+
+        val primaryCtorNode = kotlinParser.createPrimaryConstructor("(${declarationsAssignedInCtor.joinToString(", ") { it.text }})").node
+        addChild(primaryCtorNode, findChildByType(CLASS_BODY))
+        declarationsAssignedInCtor.forEach { ktProperty ->
+            ktProperty.node.let { treeParent.removeChild(it) }
+        }
+
+        if (otherStatements.isNotEmpty()) {
+            findChildByType(CLASS_BODY)?.run {
+                val beforeNode = findChildByType(LBRACE)!!.let { if (it.treeNext.elementType == WHITE_SPACE) it.treeNext else it }
+                val classInitializer = kotlinParser.createNode(
+                    """
+                    |init {
+                    |    ${otherStatements.joinToString("\n") { it.text }}
```

Comment:
I started thinking that there can be issues right now:
```
constructor(a: Int) {
    val f = F(a)
    this.a = f.foo()
}
```

how it should work?

Reply:
I think we can't reliably automate complex cases. Your example will be converted to
```kotlin
class Foo(val a: Int) {
    init {
        f = F(a)
    }
}
```
and it will break the code. In my opinion, correct form of this code should be
```kotlin
class Foo(a: Int) {
    val a: Int by lazy {
        val f = F(a)
        f.foo()
    }
}
```
but there may be other valid cases for refactoring.

Reply:
please add a Kdoc with a fixme for it